scheduler/launchd.go

```diff
@@ -0,0 +1,149 @@
+package scheduler
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path"
+
+	homedir "github.com/mitchellh/go-homedir"
+)
+
+type launchd struct{}
+
+// Enable creates the launchd script and then loads it using launchctl
+func (ld *launchd) Enable() (launchdWasLoaded bool, err error) {
+	launchdAgentFilename, err := ld.GetAgentFilename()
+	if err != nil {
+		return false, ErrModifyingLaunchd{origError: err}
+	}
+
+	launchdWasLoaded, err = ld.enable(&fileFunctionsPassthrough{}, launchdAgentFilename)
+	if err != nil {
+		return launchdWasLoaded, ErrModifyingLaunchd{origError: err}
+	}
+	return launchdWasLoaded, nil
+}
+
+// Disable unloads the launchd script, then deletes the script
+func (ld *launchd) Disable() (launchdWasUnloaded bool, err error) {
+	launchdAgentFilename, err := ld.GetAgentFilename()
+	if err != nil {
+		return false, ErrModifyingLaunchd{origError: err}
+	}
+
+	launchdWasUnloaded, err = ld.disable(&fileFunctionsPassthrough{}, launchdAgentFilename)
+	if err != nil {
+		return launchdWasUnloaded, ErrModifyingLaunchd{origError: err}
+	}
+	return launchdWasUnloaded, nil
+}
+
+func (ld *launchd) Name() string {
+	return "launchd"
+}
+
+func (ld *launchd) enable(fileFunctions fileFunctionsInterface, launchdAgentFilename string) (
+	launchdFileWasCreated bool, err error) {
+
+	if _, err := fileFunctions.OsStat(launchdAgentFilename); os.IsNotExist(err) {
+		// file does not exist, write out default launchd agent file
+		if err = fileFunctions.IoutilWriteFile(
+			launchdAgentFilename, []byte(LaunchdFileContents), 0644); err != nil {
+			return false,
+				fmt.Errorf("%s didn't exist and failed to create it: %v", launchdAgentFilename, err)
+		}
+	}
+
+	_, err = ld.runLaunchctl("load", launchdAgentFilename)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (ld *launchd) disable(fileFunctions fileFunctionsInterface, launchdAgentFilename string) (
+	launchdFileWasRemoved bool, err error) {
+
+	_, err = ld.runLaunchctl("unload", launchdAgentFilename)
+	if err != nil {
+		return false, err
+	}
+
+	if err := fileFunctions.OsRemove(launchdAgentFilename); err != nil {
+		return false, fmt.Errorf("failed to remove %s: %v", launchdAgentFilename, err)
+	}
+
+	return true, nil
+}
+
+func (ld *launchd) GetAgentFilename() (string, error) {
+	homeDirectory, err := homedir.Dir()
+	if err != nil {
+		return "", err
+	}
+
+	return path.Join(homeDirectory, LaunchAgentsSubdirectory, LaunchAgentFilename), nil
+}
+
+func (ld *launchd) runLaunchctl(verb string, filename string) (string, error) {
+	log.Printf("Running `%s %s %s`", launchctl, verb, filename)
+	cmd := exec.Command(launchctl, verb, filename)
+
+	out, err := cmd.CombinedOutput()
+
+	outString := string(out)
+	if err != nil {
+		return outString, err
+	}
+	if outString == filename+": No such file or directory\n" {
```

Comment:
This implies that if you call `launchctl <non existent file>` it returns a *successful* exit code (hence above `err` is nil`) but yet prints out "No such file or directory"... is that true?

Reply:
As far as I can tell...
```
launchctl load nonexistant.plist
/Users/idrysdale/go/src/github.com/fluidkeys/fluidkeys/nonexistant.plist: No such file or directory
➜  fluidkeys git:(add-launchd) ✗ echo $?
0
```

Reply:
Absolutely right! It's worth a comment there to document `luanchctl`'s (weird) error behaviour